src/crypto/openssl/hash.cpp

```diff
@@ -37,13 +37,73 @@ namespace crypto
 
   using namespace OpenSSL;
 
+  static thread_local EVP_MD_CTX* mdctx = nullptr;
+  static thread_local EVP_MD_CTX* basectx = nullptr;
+
+  void openssl_sha256_init()
+  {
+    if (mdctx || basectx)
+    {
+      return; // Already initialised
+    }
+
+    mdctx = EVP_MD_CTX_new();
+    if (mdctx == nullptr)
+    {
+      throw std::logic_error("openssl_sha256_init: failed to create mdctx");
+    }
+    basectx = EVP_MD_CTX_new();
+    if (basectx == nullptr)
+    {
+      throw std::logic_error("openssl_sha256_init: failed to create basectx");
+    }
+    if (EVP_DigestInit_ex(basectx, EVP_sha256(), nullptr) != 1)
+    {
+      throw std::logic_error("EVP_DigestInit_ex failed");
```

Comment:
Same question as above.

Reply:
It seems cleaner to destroy everything if we throw, and reset to init state, than to leave a partially constructed situation with some heap allocations.